### PR TITLE
Fix some minor issues in Go codes templates

### PR DESCRIPTION
**Pull request by houkms**
Fix some minor issues in Go codes templates.

### CODE REVIEW

File: templates/azure/terraform/expand_property_method.erb [hunk 1]
```diff
@@ -74,6 +74,7 @@ func expand<%= sdk_marshal.resource -%><%= descriptor.func_name -%>(input <%= go
 <%        elsif property.is_a?(Api::Type::Array) -%>
     results := make([]<%= sdk_marshal.package -%>.<%= sdk_marshal.sdktype.go_type_name -%>, 0)
     for _, v := range input {
+      v := v.(map[string]interface{})
```
Comment (JunyiYi) [inline templates/azure/terraform/expand_property_method.erb]:
```suggestion
        v := item.(map[string]interface{})
```

File: templates/azure/terraform/expand_property_method.erb [hunk 2]
```diff
@@ -74,6 +74,7 @@ func expand<%= sdk_marshal.resource -%><%= descriptor.func_name -%>(input <%= go
 <%        elsif property.is_a?(Api::Type::Array) -%>
     results := make([]<%= sdk_marshal.package -%>.<%= sdk_marshal.sdktype.go_type_name -%>, 0)
     for _, v := range input {
```
Comment (JunyiYi) [inline templates/azure/terraform/expand_property_method.erb]:
```suggestion
     for _, item := range input {
```
And rename the original `item` to `result`.

**Review by JunyiYi — CHANGES_REQUESTED** [review]
Thanks @houkms for the update. The code looks good except for some tiny naming and space issues. Let's use this PR to track most of the issues discovered in the Azure Frontdoor Firewall Policy resource.

File: templates/azure/terraform/expand_property_method.erb (hunk 1 above)
Comment (houkms) [reply templates/azure/terraform/expand_property_method.erb]: Should it be `item := item.(map[string]interface{})` ?

The generated .go code is like [this](https://github.com/terraform-providers/terraform-provider-azurerm/compare/master...houkms:houk-dev#diff-072322d5bd7284c81dca26fd0afae3f3R348-R349)

File: templates/azure/terraform/expand_property_method.erb (hunk 2 above)
Comment (houkms) [reply templates/azure/terraform/expand_property_method.erb]: ok.

**Comment by houkms** [comment]
@JunyiYi Codes are modified, find your time to check it.

File: templates/azure/terraform/expand_property_method.erb (hunk 1 above)
Comment (JunyiYi) [reply templates/azure/terraform/expand_property_method.erb]: No, it should be `v := item.(map[string]interface{})`

**Comment by JunyiYi** [comment]
Thanks @houkms for the update. Please use `v := item.(map[string]interface{})` instead of `item := ...` in the template because we will use `v` later in `build_schema_property_get('v', output_var, prop, sdk_marshal, 8)`.

**Comment by houkms** [comment]
OK, fixed.

**Comment by houkms** [comment]
This PR fix the issue in this magic-module-specs [PR](https://github.com/Azure/magic-module-specs/pull/84).

**Comment by houkms** [comment]
@JunyiYi  Other possible issues are fixed in the above commit, including

- Go SDK rename: FlattenStringArray --> FlattenStringSlice and ExpandStringArray --> ExpandStringSlice
- String array decision condition adjustment
- Terraform configuration grammer: use block for array
- Introduce help function to flatten example properties

Please find your time to review.

**Comment by houkms** [comment]
This PR also fixes the pipeline error in the magic-module-specs [PR102](https://github.com/Azure/magic-module-specs/pull/102)

File: provider/azure/terraform/example/helpers.rb [hunk 3]
```diff
@@ -11,13 +11,37 @@ def get_example_properties_to_check(example_name, object)
...
+          end
+
+          def flatten_example_properties_to_check(properties, is_nested)
+            return properties unless is_nested
+            flat_properties = Hash.new
+            next_is_nested = false
+            properties.each do |pn, pv|
+              if pv.is_a?(Hash)
+                pv.each do |key, val|
+                  flat_properties[pn+"."+key] = val
```
Comment (JunyiYi) [inline provider/azure/terraform/example/helpers.rb]: You can simplify the string with `"#{pn}.#{key}"`.

File: provider/azure/terraform/example/helpers.rb [hunk 4]
```diff
@@ -11,13 +11,37 @@ def get_example_properties_to_check(example_name, object)
...
               .reject do |pn, pv|
                 params.include?(pn) || pn == 'location'
               end
               .transform_values do |v|
                 v.is_a?(String) && !v.match(/\$\{.+\}/).nil? ? :AttrSet : v
               end
+            flatten_example_properties_to_check(example_props, true)
+          end
+
+          def flatten_example_properties_to_check(properties, is_nested)
```
Comment (JunyiYi) [inline provider/azure/terraform/example/helpers.rb]: Here `is_nested`, do you mean `has_nested_item`?

File: templates/azure/terraform/example/hcl_properties.erb [hunk 5]
```diff
@@ -12,6 +12,9 @@
 <%=       prop[0].ljust(prop_alignment) -%> = "<%= prop[1] -%>"
 <%      elsif prop[1].is_a?(Integer) || prop[1].is_a?(TrueClass) || prop[1].is_a?(FalseClass) -%>
 <%=       prop[0].ljust(prop_alignment) -%> = <%= lines(prop[1].to_s) -%>
+<%      elsif prop[1].is_a?(Array) && (prop[1].length == 0 || prop[1][0].is_a?(String))
+          res_arr = prop[1].collect{|elem| "\""+elem.to_s+"\""} -%>
```
Comment (JunyiYi) [inline templates/azure/terraform/example/hcl_properties.erb]: Can you try the following code whether it works or not?

```suggestion
          res_arr = prop[1].map{|elem| "\"#{elem}\""} -%>
```

**Review by JunyiYi — CHANGES_REQUESTED** [review]
Thanks @houkms , after taking a look at the PR, I think most of the string concatenation or `to_s` stuffs could be simplified by [Ruby String Interpolation](http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html).

File: provider/azure/terraform/example/helpers.rb (hunk 3 above)
Comment (houkms) [reply provider/azure/terraform/example/helpers.rb]: Replaced! Also done for other similar cases.

File: provider/azure/terraform/example/helpers.rb (hunk 4 above)
Comment (houkms) [reply provider/azure/terraform/example/helpers.rb]: Yes, and the name was changed to be `has_nested_item`

File: templates/azure/terraform/example/hcl_properties.erb (hunk 5 above)
Comment (houkms) [reply templates/azure/terraform/example/hcl_properties.erb]: It works, replaced now.

**Comment by houkms** [comment]
@JunyiYi Codes were refined now according to your suggestions.

File: templates/azure/terraform/example/hcl_properties.erb [hunk 6]
```diff
@@ -24,7 +27,7 @@
         blocks.each do |block|
 -%>
 
-<%=       prop[0] -%> = {
+<%=       prop[0] -%> {
```
Comment (JunyiYi) [inline templates/azure/terraform/example/hcl_properties.erb]: Please make sure we include the equal sign (`=`) for tags (You can define a new helper function in `terraform/helpers.rb` similar to [this one for ansible](https://github.com/VSChina/magic-modules/blob/azure_backend/provider/azure/ansible/helpers.rb#L24-L26)). Here is the example in Terraform: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_eventgrid_topic_test.go#L201.

**Review by JunyiYi — CHANGES_REQUESTED** [review]
The code is pretty good to go except for one minor issue. We still need to keep the `=` for tags while omit it for other block properties.

**Comment by houkms** [comment]
Added codes to deal with the special case for tags property.

File: templates/azure/terraform/example/hcl_properties.erb [hunk 7]
```diff
@@ -24,7 +27,7 @@
         blocks.each do |block|
 -%>
 
-<%=       prop[0] -%> = {
+<%=       prop[0] -%> <%= "= " if is_tags?(prop)-%>{
```
Comment (JunyiYi) [inline templates/azure/terraform/example/hcl_properties.erb]: Well, I think for this case, it is not a good choice to introduce that common helper function. Let's do it inline: `<%= '= ' if prop[0] == 'tags' -%>`.

**Review by JunyiYi — APPROVED** [review]
Code LGTM, let's merge that in.

**Pull request merged**